microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/ConfigSourceDefinition.java

```diff
@@ -166,4 +130,60 @@ private static Class unwrapClass(ModelNode classModel) throws OperationFailedExc
             throw ROOT_LOGGER.unableToLoadClassFromModule(className, moduleName);
         }
     }
+
+    private static class ConfigSourceDefinitionAddHandler extends AbstractAddStepHandler {
+        private final Registry<ConfigSource> sources;
+
+        ConfigSourceDefinitionAddHandler(Registry<ConfigSource> sources) {
+            super(ATTRIBUTES);
+            this.sources = sources;
+        }
+
+        @Override
+        protected void performRuntime(OperationContext context, ModelNode operation, ModelNode model) throws OperationFailedException {
+            super.performRuntime(context, operation, model);
+            String name = context.getCurrentAddressValue();
+            int ordinal = ORDINAL.resolveModelAttribute(context, model).asInt();
+            ModelNode props = PROPERTIES.resolveModelAttribute(context, model);
+            ModelNode classModel = CLASS.resolveModelAttribute(context, model);
+            ModelNode dirModel = DIR.resolveModelAttribute(context, model);
+            if (classModel.isDefined()) {
+                Class configSourceClass = unwrapClass(classModel);
+                try {
+                    sources.register(name, ConfigSource.class.cast(configSourceClass.newInstance()));
+                    MicroProfileConfigLogger.ROOT_LOGGER.loadConfigSourceFromClass(configSourceClass);
+                } catch (Exception e) {
+                    throw new OperationFailedException(e);
+                }
+            } else if (dirModel.isDefined()) {
+                String path = PATH.resolveModelAttribute(context, dirModel).asString();
+                String relativeTo = RELATIVE_TO.resolveModelAttribute(context, dirModel).asStringOrNull();
+                DirConfigSourceRegistrationService.install(context, name, path, relativeTo, ordinal, sources);
+            } else {
+                Map<String, String> properties = PropertiesAttributeDefinition.unwrapModel(context, props);
+                sources.register(name, new PropertiesConfigSource(properties, name, ordinal));
+            }
+        }
+    }
+
+    private static class ConfigSourceDefinitionRemoveHandler extends ReloadRequiredRemoveStepHandler {
+        private final Registry<ConfigSource> sources;
+
+        private ConfigSourceDefinitionRemoveHandler(Registry<ConfigSource> sources) {
+            this.sources = sources;
+        }
+
+        @Override
+        protected void performRuntime(OperationContext context, ModelNode operation, ModelNode model) throws OperationFailedException {
+            String name = context.getCurrentAddressValue();
+            ModelNode dirModel = DIR.resolveModelAttribute(context, model);
+            if (dirModel.isDefined()) {
+                context.removeService(ServiceNames.CONFIG_SOURCE.append(name));
+            } else {
+                sources.unregister(name);
+            }
+            super.performRuntime(context, operation, model);
+        }
```

Comment:
@jasondlee This should just be ReloadRequiredRemoveStepHandler, no override.

A management op handler should do one of 4 mutually exclusive things with respect to the running behavior of a server:

1) Fail because something is wrong.
2) Determine it can do what was requested immediately and go ahead and change the runtime services in some way to bring their state in sync with the current configuration.
3) Determine that it can't do what's requested without a reload and therefore put the server in reload-required.
4) Determine that it can't do what's requested without process restart and therefore put the server in restart-required.

It shouldn't partially change the running services but then say reload/restart are required to finish the job.